File: frontend/src/app/core/services/cuenta-bancaria.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  CuentaBancaria,
  CreateCuentaBancariaDto,
  UpdateCuentaBancariaDto,
  CuentaBancariaQueryFilter
} from '../models/cuenta-bancaria';
import { CuentaBancariaPage } from '../models/cuenta-bancaria-page';

@Injectable({
  providedIn: 'root'
})
export class CuentaBancariaService {
  private readonly apiUrl = `${environment.apiUrl}/cuentas-bancarias`;

  constructor(private readonly http: HttpClient) {}

  getAll(filter?: CuentaBancariaQueryFilter): Observable<CuentaBancariaPage<CuentaBancaria>> {
    let params = new HttpParams();

    if (filter) {
      if (filter.page != null) params = params.set('page', filter.page.toString());
      if (filter.pageSize != null) params = params.set('pageSize', filter.pageSize.toString());
      if (filter.bancoId != null) params = params.set('bancoId', filter.bancoId.toString());
      if (filter.estado != null) params = params.set('estado', filter.estado.toString());
      if (filter.moneda != null) params = params.set('moneda', filter.moneda);
      if (filter.searchTerm != null) params = params.set('searchTerm', filter.searchTerm);
    }

    return this.http.get<CuentaBancariaPage<CuentaBancaria>>(this.apiUrl, { params });
  }

  getActivas(): Observable<CuentaBancaria[]> {
    return this.http.get<CuentaBancaria[]>(`${this.apiUrl}/activas`);
  }

  getById(id: number): Observable<CuentaBancaria> {
    return this.http.get<CuentaBancaria>(`${this.apiUrl}/${id}`);
  }

  create(dto: CreateCuentaBancariaDto): Observable<CuentaBancaria> {
    return this.http.post<CuentaBancaria>(this.apiUrl, dto);
  }

  update(id: number, dto: UpdateCuentaBancariaDto): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${id}`, dto);
  }

  activar(id: number): Observable<void> {
    return this.http.patch<void>(`${this.apiUrl}/${id}/activar`, {});
  }

  desactivar(id: number): Observable<void> {
    return this.http.patch<void>(`${this.apiUrl}/${id}/desactivar`, {});
  }
}

File: backend/src/Application/Services/CuentaBancariaService.cs
using InventoryApp.Application.DTOs.Bancos;
using InventoryApp.Application.Interfaces;
using InventoryApp.Domain.Entities.Bancos;
using InventoryApp.Application.Bancos;
using InventoryApp.Domain.Enums;

namespace InventoryApp.Application.Services;

public sealed class CuentaBancariaService : ICuentaBancariaService
{
    private readonly ICuentaBancariaRepository _repository;
    private readonly IAuditoriaService _auditoria;

    public CuentaBancariaService(ICuentaBancariaRepository repository, IAuditoriaService auditoria)
    {
        _repository = repository;
        _auditoria = auditoria;
    }

    public async Task<CuentaBancariaDto?> GetByIdAsync(int id)
    {
        var cuenta = await _repository.GetByIdAsync(id);
        if (cuenta == null) return null;

        return MapToDto(cuenta);
    }

    public async Task<CuentaBancariaPage<CuentaBancariaDto>> GetAllAsync(CuentaBancariaQueryFilter filter)
    {
        var page = await _repository.GetAllAsync(filter);
        var dtoItems = page.Items.Select(MapToDto).ToList();
        return new CuentaBancariaPage<CuentaBancariaDto>(dtoItems, page.Page, page.PageSize, page.TotalCount);
    }

    public async Task<List<CuentaBancariaDto>> GetActivasAsync()
    {
        var cuentas = await _repository.GetActivasAsync();
        return cuentas.Select(MapToDto).ToList();
    }

    public async Task<CuentaBancariaDto> AddAsync(CreateCuentaBancariaDto dto)
    {
        var cuenta = new CuentaBancaria(
            dto.BancoId,
            dto.Nombre,
            dto.NumeroCuenta,
            dto.Moneda,
            dto.SaldoInicial);

        await _repository.AddAsync(cuenta);
        await _repository.SaveChangesAsync();

        await _auditoria.RegistrarAsync(
            ModuloSistema.Finanzas,
            AccionPermiso.Crear,
            $"Creó la cuenta bancaria '{cuenta.Nombre}'.",
            cuenta.Id,
            entidad: "CuentaBancaria",
            valoresNuevos: new { cuenta.BancoId, cuenta.Nombre, cuenta.Moneda, cuenta.SaldoInicial, cuenta.Estado });

        return MapToDto(cuenta);
    }

    public async Task ActivarAsync(int id)
    {
        var cuenta = await _repository.GetByIdAsync(id)
            ?? throw new InvalidOperationException($"No se encontró la cuenta con Id {id}.");

        cuenta.Activar();
        _repository.Update(cuenta);
        await _repository.SaveChangesAsync();

        await _auditoria.RegistrarAsync(
            ModuloSistema.Finanzas,
            AccionPermiso.Activar,
            $"Activó la cuenta bancaria '{cuenta.Nombre}'.",
            cuenta.Id,
            entidad: "CuentaBancaria");
    }

    public async Task DesactivarAsync(int id)
    {
        var cuenta = await _repository.GetByIdAsync(id)
            ?? throw new InvalidOperationException($"No se encontró la cuenta con Id {id}.");

        cuenta.Desactivar();
        _repository.Update(cuenta);
        await _repository.SaveChangesAsync();

        await _auditoria.RegistrarAsync(
            ModuloSistema.Finanzas,
            AccionPermiso.Desactivar,
            $"Desactivó la cuenta bancaria '{cuenta.Nombre}'.",
            cuenta.Id,
            entidad: "CuentaBancaria");
    }

    public async Task UpdateAsync(int id, UpdateCuentaBancariaDto dto)
    {
        var cuenta = await _repository.GetByIdAsync(id)
            ?? throw new InvalidOperationException($"No se encontró la cuenta con Id {id}.");

        var nombreAnterior = cuenta.Nombre;
        cuenta.UpdateNombre(dto.Nombre);
        _repository.Update(cuenta);
        await _repository.SaveChangesAsync();

        await _auditoria.RegistrarAsync(
            ModuloSistema.Finanzas,
            AccionPermiso.Editar,
            $"Editó la cuenta bancaria '{cuenta.Nombre}'.",
            cuenta.Id,
            entidad: "CuentaBancaria",
            valoresAnteriores: new { Nombre = nombreAnterior },
            valoresNuevos: new { cuenta.Nombre });
    }

    private static CuentaBancariaDto MapToDto(CuentaBancaria cuenta) => new()
    {
        Id = cuenta.Id,
        BancoId = cuenta.BancoId,
        Nombre = cuenta.Nombre,
        NumeroCuenta = cuenta.NumeroCuenta,
        Moneda = cuenta.Moneda,
        SaldoInicial = cuenta.SaldoInicial,
        Estado = cuenta.Estado
    };
}
